### DISCUSSION TITLE
2023/03/20 프로토타입 개발을 위한 준비

### DISCUSSION THREAD

**Discussion by hyunnnn98** [2P3S/scrum-dice#15] [General]
# ⏰ 투두리스트

- [ ] 승현 - 프로토타이핑 도구로 유저 스토리 만들어보기
- [ ] 재순 - 서비스 기능 정리 + 승형 서포트
- [ ] 승형 - Next.js 채팅 어플리케이션 만들기 + Vercel 배포 후 기술스택 검증

**Comment by hyunnnn98** [comment]
# 🎱 프로토타입 디자인
- 프로토타이핑 도구는 [Figma](https://www.figma.com/files/recent?fuid=982444716018702393)
- [Wireframes UI Kit](https://www.figma.com/community/file/1209724056553648259)를 이용하여 와이어를 배치하고 있습니다
- 각 컴포넌트 안에 들어갈 Text는 비워놓기보다는 임의로 채워 넣었습니다. 추후 상의해서 매꿔가는 식으로 해봐요~
### 홈페이지
![CleanShot 2023-03-16 at 00 07 58](https://user-images.githubusercontent.com/52916934/225352903-d6253e09-5055-4822-bfb1-5f675cd255d5.png)
### 비회원 로그인 페이지
I'll Be Back Soon
### room 생성 페이지
I'll Be Back Soon
### playground 페이지
I'll Be Back Soon

**Reply by JeongJaeSoon** [reply]
필요한 이미지 생성에 [마이크로소프트 디자이너](https://designer.microsoft.com/ )를 활용해봐도 재미있을 것 같음
→ 활용 후기：https://www.itworld.co.kr/news/267495、https://brunch.co.kr/@seungpillee/55

**Reply by JeongJaeSoon** [reply]
https://www.canva.com/ko_kr/features/ai-image-generator/
↑ 다양한 AI 이미지 생성기도 활용 가능

**Reply by JeongJaeSoon** [reply]
Text 에 대해서도 Chat GPT 의 도움을 받는다면, 이후 개발 스토리로서 재미있는 요소가 될 수도

**Reply by hyunnnn98** [reply]
https://www.figma.com/file/rhYzh5jurgqhfMt1m6wDT3/Scrum-Dice?node-id=0%3A1&t=Gz9LgezVeG3cDKOH-1

**Comment by JeongJaeSoon** [comment]
# 서비스 기능

### 1. 사용자 화면 진입

- 메인화면
  - 플래닝 포커 설명?
- 로그인
  - Guest 로그인 (must)
  - SSO 연동 로그인 (향후)

**Comment by JeongJaeSoon** [comment]
### 2. 플래닝 포커 룸 생성

- 룸 이름 + 패스워드 설정 (must)
  - UID로 각 방을 구분
  - 이름 자체는 중복 가능
  - 패스워드는 자동 생성 (URL 에 hash 처리)
  - 룸 생성 시 자동으로 공유용 URL 클립보드 복사
- 룸 이름 미입력 시, Default 로 날짜 (must)
  - user_name's room(2012-34-56)
- 유효시간 내에 입장 기록이 있을 경우, 재입장 가능 (must)
  - 같은 방에 같은 이름? 아니면 같은 사람으로 추측되었을 때
  - 전에 입장한 기록이 있네요? 00 로 입장하시겠습니까?
  - URL 로 접근했을 때, 연결이 끊어진 사람 목록을 보여주면서 혹시 이 사용자인가요?

**Comment by JeongJaeSoon** [comment]
### 3. 플래닝 포커 진행

- 사용 기록 없을 경우, 플래닝 포커 설명 노출 (향후)
  - 데모처럼 사용설명 나오는 튜토리얼 형식
- 플래닝 포커의 흐름은 다음과 같음
  1. 회차 이름 설정(Issue number) or Default 는 00 회차
  2. Egg timer 시작 (선택)
  3. 아이템에 대하여, 담당자가 설명
  4. 모두가 포커 카드를 선택
  5. 포커 카드를 뒤집음 (자동 평균 계산)
- 포커 구성
  - 기본 숫자 카드 (must)
    - 1안 : 0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89
    - 2안 : 0, ½, 1, 2, 3, 5, 8, 13, 20, 40, 100
    - 룸 생성 시, 2가지 카드 묶음 중 하나를 선택
  - 그 외 카드 (must)
    - ☕️ Coffee : 이번 턴은 쉬고 싶음
    - ? : 잘 모르겠음
    - 👑 King : 산정하기에 너무 덩치가 크다 or 또는 너무 복잡하다
- Egg timer (want)
  - 하나의 아이템에 관해 토론할 수 있는 시간
  - 시간이 다 되었을 때, 토론이 중지되고 포커를 낼 수 있어야 함
- 두 사람이 이상이 선택했는데, 내가 아직 선택하지 않았을 경우 Chime bell (향후)
- 숫자 차이가 클 경우 강조 표시가 됨 (must)

  - 표준편차의 상/하위를 걸러내서 표시

    ```javascript
    // let array = [3, 1, 3, 5, 13, 3, 3, 1, 3, 5, 13, 3];
    // let array = [3, 1, 3, 5, 3, 3, 3, 1, 3, 5, 1, 3];
    // let array = [3, 1, 3, 3, 3, 3, 3, 1, 3, 3, 1, 3];
    // let array = [3, 1, 3, 5, 3, 3, 3, 1, 3, 5, 5, 3];

    function myFunction(values) {
      let sum = 0;
      let n = values.length;
      for (let i = 0; i < n; i++) sum = sum + values[i];

      return sum / n;
    }

    function myAverage(arr) {
      return arr.reduce((p, c) => p + c, 0) / arr.length;
    }

    let result = [];
    let rate = parseInt(array.length / 5);
    let median = myFunction(array);
    let average = myAverage(array);

    console.log(`median : ${median}`);

    array.map((ele, idx) => {
      let diff = ele - median;
      result.push({ ele, diff });
      // console.log(`d${ele} 편차 : ${diff}`);
    });
    result.sort((a, b) => {
      return a.diff - b.diff;
    });

    if (Math.abs(result[0].diff - result[result.length - 1].diff) > 3) {
      console.log(result.slice(0, rate));
      console.log(result.slice(result.length - rate));
    }
    ```

  - outlier detected
    - https://stackoverflow.com/questions/20811131/javascript-remove-outlier-from-an-array
    - https://www.npmjs.com/package/outliers

**Comment by JeongJaeSoon** [comment]
### 4. 플래닝 포커 종료

- 기록 보존
  - 회차별 결과 테이블로 확인 가능 (must)
  - URL 로 생성, 일정기간 보존 (향후)
    - Redis, ~~DB~~ 등을 사용해서 유지
  - "내 기록 보존 / 조회"은 SSO 로 로그인한 유저만 가능 (DB) (향후)
- Shorts URL 생성 기능 필요 (향후)
  - 구현방법?

**Comment by JeongJaeSoon** [comment]
# 그 외 부가 기능

- Chrome Extensions 으로 바로 방 생성

**Comment by Samsam-lee** [comment]
📝 2023/03/20 회의록
---
### UI
- https://www.figma.com/file/rhYzh5jurgqhfMt1m6wDT3/Scrum-Dice?node-id=0%3A1&t=Gz9LgezVeG3cDKOH-1
- 페이지 용어적인 부분에서 사용자들이 알기 쉽게 보여주는 도큐먼트가 있으면 좋겠음
- Microsoft Designer / Canva 같은 것들 사용하면 좋을 것 같다.

### 다국어
- 필요성은 있다고 생각함 (한국어, 영어, 일본어, etc...)
- UI 생각하지 않고 개발적인 측면만 생각해도 좋을 것 같다.

### 기능
- 사용자 화면 진입
    - 게스트 로그인 (must)
    - SSO 로그인 (추후 기록과 같은 데이터 보여줄 때 필요)
- 플래닝 포커 룸 생성
    - 룸 이름 + 패스워드 설정 (must)
    - 룸 이름 미입력 시 default -> 날짜
    - 유효시간 내에 입장 기록이 있을 경우, 재입장 가능 (must)
- 플래닝 포커 진행
    - 숫자 카드 1, 2안..
    - 커피, 물음표 카드
    - 숫자 차이가 클 경우 강조 표시가 됨 (조금 더 생각해보기)
    - ✔️ 큰 수, 작은 수?
    - ✔️ 표준편차?
- 플래닝 포커 종료
    - 회차별 결과 테이블로 확인 가능 (must)
    - URL 로 생성, 일정기간 보존 (향후)
    - "내 기록 보존 / 조회"은 SSO 로 로그인한 유저만 가능 (DB) (향후)
    - Shorts URL 생성 기능 필요 (향후)

### 다음시간
- [ ] 승현 : Figma UI(메인 로직 대략적인 스토리), Front-End tech stack 검토
- [ ] 승형 : 소켓 서버 구현 후 배포
- [ ] 재순 : 소켓 로직 설계

- 낮 시간에는 디코로 대화 / 질문
- 꼬실 사람? ㅎ (6월 예정)
